homeassistant/components/switch/mochad.py

```diff
@@ -50,24 +50,46 @@ def __init__(self, hass, ctrl, dev):
         self._comm_type = dev.get(mochad.CONF_COMM_TYPE, 'pl')
         self.device = device.Device(ctrl, self._address,
                                     comm_type=self._comm_type)
-        self._state = self._get_device_status()
+        """Init with false to avoid locking HA for long on CM19A"""
+        self._state = False
+        self.lock = threading.Lock()
 
     @property
     def name(self):
         """Get the name of the switch."""
         return self._name
 
     def turn_on(self, **kwargs):
-        """Turn the switch on."""
-        self._state = True
-        self.device.send_cmd('on')
-        self._controller.read_data()
+        try:
+            self.lock.acquire()
+            """Turn the switch on."""
+            self._state = True
+            """Recycle socket on new command to recover mochad connection"""
+            _LOGGER.debug("Reconnect {} : {} ".format(self._controller.server, 
+                                                      self._controller.port))
+            self._controller.reconnect()
+            self.device.send_cmd('on')
+            self._controller.read_data()
+        except Exception as e:
+            _LOGGER.error("Error with mochad communication: {}".format(e))
+        finally:
+            self.lock.release()
 
     def turn_off(self, **kwargs):
-        """Turn the switch off."""
-        self._state = False
-        self.device.send_cmd('off')
-        self._controller.read_data()
+        try:
+            self.lock.acquire()
```

Comment:
Use `with` blocks to acquire/release locks.

```python
with self.lock:
    try:
        …
```

Reply:
Ok. fixed in commit 3. e88301b